daemons/controld/controld_execd.c

```diff
@@ -594,9 +598,11 @@ append_secure_list(lrmd_event_data_t *op, struct ra_metadata_s *metadata,
      * the insecure ones
      */
     secure = create_xml_node(NULL, XML_TAG_PARAMS);
+    g_hash_table_foreach(op->params, hash2field, secure);
```

Comment:
I'm unclear why this would need to be done. Any action-specific instance attributes would already have been incorporated in the parameters the controller gets via the graph. I'd rather not change the controller end if we don't have to, so we don't need to think about rolling upgrades and what not ...

Reply:
This is about recording a correct secure digest for future use. Technically a secure digest recorded by controller will be sensible for scheduler/crm_simulate to check against later only if it's calculated the consistent way,

So far this is mainly for preventing unreasonable inconsistencies between scheduler and crm_simulate. Considering a reload/start performed with op-specific parameters got recorded but with a secure digest without taking the op-specific parameters into account. If there's any further changes of  op-specific parameters,  in practice reload will be triggered again, which is fine. But if we check the transition history with crm_simulate, crm_simulate cannot reproduce the same transition due to mistakenly thinking "Only 'private' parameters" changed.

But indeed, it depends on if it's worth it to change the behavior of controller now to prevent such confusions ...

Reply:
But how could the controller have ever recorded a secure digest without op-specific parameters? The controller generates its hash based on the parameters provided in the transition graph, which the scheduler has always generated with op-specific parameters taken into account. It's just the scheduler's own hash for comparison that's affected by not including op-specific parameters. Unless I'm missing something ...

Reply:
The invocation of `build_parameter_list()` below builds up the `secure` xml only with the parameters that are listed in the metadata of the RA without considering anything else.  And `secure` xml is what the op-secure-digest is based on, which will be appended into lrm_rsc_op and get injected into the cib.

Reply:
> But how could the controller have ever recorded a secure digest without op-specific parameters? The controller generates its hash based on the parameters provided in the transition graph, 

I think it is the case only for the main  `op-digest`.